text/0000-dyn-assoc.md

```diff
@@ -0,0 +1,371 @@
+```rust
+trait Thing {
+    fn make_a_friend(&self) -> Box<dyn Thing>;
+}
+struct SimpleThing;
+impl Thing for SimpleThing {
+    fn make_a_friend(&self) -> Box<dyn Thing> {
+        Box::from(SimpleThing)
+    }
+}
+struct BetterThing;
+impl Thing for BetterThing {
+    fn make_a_friend(&self) -> Box<dyn Thing> {
+        Box::from(BetterThing)
+    }
+}
+
+fn main() {
+    let sithing: Box<dyn Thing> = Box::from(SimpleThing);
+    let bething: Box<dyn Thing> = Box::from(BetterThing);
+    let sithing2 = sithing.make_a_friend();
+    let bething2 = sithing.make_a_friend();
+}
+```
+Supporting the actual `Self` type [might still be introduced in a later RFC][future-possibilities].
+
+## Generic contexts
+As [described][generic-contexts] in the guide-level explanation, we can put trait objects themselves in the generics to get associated members to work in functions which do not take the trait object as an argument and are not dynamically dispatched from the side which has the trait object. How does this work?
+
+For functions with generic arguments, the vtable (not the trait object, only its vtable) is passed as an invisible argument on the stack/registers, which is present only in the morphomization of the function which takes a trait object in its generics. If a function has more than one `?Sized` generic argument, the morphomization (generated at compile time) which requires multiple trait objects should take multiple invisible vtable arguments. The arguments are automatically filled out when the callee puts  the trait object(s) into the angle brackets (if they don't provide all the trait objects required and fill some with `dyn Trait`, an error is produced as described back in the guide-level section).
+
+For types with generic arguments, the vtables exist inside instances of that type as invisible private fields. Like the vtable function arguments, these are only visible to the compiler (but they obviously contribute to the size, which remains known since vtable pointers are just pointers). If an associated function is called, however (no `self` — no access to the invisible vtable fields), the vtables should be provided as invisible vtable arguments, as with regular functions.
+
+## Edge cases
+None unresolved yet.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None known yet.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The constantly mentioned alternative is declaring the associated functions in traits which are meant to be trait objects with a `&self` parameter, effectively converting all associated functions into trait methods. This becomes a major source of confusion if the trait object is mutably borrowed elsewhere while this "`&self`ified" method is called. This introduces lots of unnecessary artificial curly-bracket scopes which hurt readability in most cases.
```

Comment:
One way around this is through arbitrary self-types. Take a raw pointer as the receiver (although this isn't supported just yet, it could be). This way it is illegal to try and access the raw pointer (because it could be invalid), but even raw pointers have a valid vtable, so you can rely on that to dispatch on the required parts.

This way you must construct at least a raw pointer to access associated functions/constants.

This in combo with rust-lang/rust#64490 would allow you to ignore the borrow checker for these sorts of associated functions/constants.